crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -100,55 +244,341 @@ fn site_packages_dir_from_sys_prefix(
+/// Abstraction for a Python virtual environment.
 ///
-/// TODO: Currently we only ever return 1 path from this function:
-/// the `site-packages` directory that is actually inside the virtual environment.
-/// Some `site-packages` directories are able to also access system `site-packages` and
-/// user `site-packages`, however.
-pub fn site_packages_dirs_of_venv(
-    venv_path: &SystemPath,
-    system: &dyn System,
-) -> Result<Vec<SystemPathBuf>, SitePackagesDiscoveryError> {
-    Ok(vec![site_packages_dir_from_sys_prefix(venv_path, system)?])
+/// Most of this information is derived from the virtual environment's `pyvenv.cfg` file.
+/// The format of this file is not defined anyway, and exactly which keys are present
+/// depends on the tool that was used to create the virtual environment.
+#[derive(Debug)]
+pub struct VirtualEnvironment {
+    venv_path: SysPrefixPath,
+    base_executable_home_path: PythonHomePath,
+    include_system_site_packages: bool,
+
+    /// The minor version of the Python executable that was used to create this virtual environment.
+    ///
+    /// The Python version is encoded under different keys and in different formats
+    /// by different virtual-environment creation tools,
+    /// and the key is never read by the standard-library `site.py` module,
+    /// so it's possible that we might not be able to find this information
+    /// in an acceptable format under any of the keys we expect.
+    /// This field will be `None` if so.
+    minor_version: Option<PythonMinorVersion>,
+}
+
+impl VirtualEnvironment {
+    pub fn new(
+        path: impl Into<SystemPathBuf>,
+        system: &dyn System,
+    ) -> Result<Self, SitePackagesDiscoveryError> {
+        Self::new_impl(path.into(), system)
+    }
+
+    fn new_impl(
+        path: SystemPathBuf,
+        system: &dyn System,
+    ) -> Result<Self, SitePackagesDiscoveryError> {
+        let Some(venv_path) = SysPrefixPath::new(&path, system) else {
+            return Err(SitePackagesDiscoveryError::VenvDirIsNotADirectory(path));
+        };
+
+        let pyvenv_cfg_path = venv_path.join("pyvenv.cfg");
+        tracing::debug!("Attempting to parse virtual environment metadata at {pyvenv_cfg_path}");
+
+        let pyvenv_cfg = system
+            .read_to_string(&pyvenv_cfg_path)
+            .map_err(SitePackagesDiscoveryError::NoPyvenvCfgFile)?;
+
+        let mut include_system_site_packages = false;
+        let mut base_executable_home_path = None;
+        let mut version_info_string = None;
+
+        // A `pyvenv.cfg` file *looks* like a `.ini` file, but actually isn't valid `.ini` syntax!
+        // The Python standard-library's `site` module parses these files by splitting each line on
+        // '=' characters, so that's what we should do as well.
+        //
+        // See also: https://snarky.ca/how-virtual-environments-work/
+        for (index, line) in pyvenv_cfg.lines().enumerate() {
```

Comment:
Do we need to worry about env files that use `\r` file endings? I'm asking because `lines` only supports `\r\n` and `\n`.

Reply:
How do we handle this elsewhere in Ruff? Is there an easy way of iterating line by line that would be resilient to files that have `\r` endings?

My instinct would be to say that there's a 95% chance that we'd never encounter a `pyvenv.cfg` file with `\r` endings, so I'm inclined not to worry about this unless there's an easy fix

Reply:
95% is kind of low haha. There's the `UniversalNewlinesIterator` in `ruff_source_file`.

Reply:
I'm running into some annoying lifetime issues using `universal_newlines`. I tried making this change:

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 970f8db7f..30aeed0c8 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1961,6 +1961,7 @@ dependencies = [
  "ruff_cache",
  "ruff_db",
  "ruff_python_ast",
+ "ruff_source_file",
  "rustc-hash 2.0.0",
  "salsa",
  "thiserror",
diff --git a/crates/red_knot_workspace/Cargo.toml b/crates/red_knot_workspace/Cargo.toml
index d8d5203f6..dbe031c29 100644
--- a/crates/red_knot_workspace/Cargo.toml
+++ b/crates/red_knot_workspace/Cargo.toml
@@ -17,6 +17,7 @@ red_knot_python_semantic = { workspace = true }
 ruff_cache = { workspace = true }
 ruff_db = { workspace = true, features = ["os", "cache"] }
 ruff_python_ast = { workspace = true }
+ruff_source_file = { workspace = true }
 
 anyhow = { workspace = true }
 crossbeam = { workspace = true }
diff --git a/crates/red_knot_workspace/src/site_packages.rs b/crates/red_knot_workspace/src/site_packages.rs
index 575d7f69a..0c9462ae9 100644
--- a/crates/red_knot_workspace/src/site_packages.rs
+++ b/crates/red_knot_workspace/src/site_packages.rs
@@ -15,6 +15,7 @@ use std::ops::Deref;
 
 use red_knot_python_semantic::PythonVersion;
 use ruff_db::system::{System, SystemPath, SystemPathBuf};
+use ruff_source_file::UniversalNewlines;
 
 type SitePackagesDiscoveryResult<T> = Result<T, SitePackagesDiscoveryError>;
 
@@ -70,7 +71,7 @@ impl VirtualEnvironment {
         // '=' characters, so that's what we should do as well.
         //
         // See also: https://snarky.ca/how-virtual-environments-work/
-        for (index, line) in pyvenv_cfg.lines().enumerate() {
+        for (index, line) in pyvenv_cfg.universal_newlines().enumerate() {
             if let Some((key, value)) = line.split_once('=') {
                 let key = key.trim();
                 if key.is_empty() 
```

But now the borrow checker complains:

```
error[E0597]: `line` does not live long enough
   --> crates/red_knot_workspace/src/site_packages.rs:75:41
    |
74  |         for (index, line) in pyvenv_cfg.universal_newlines().enumerate() {
    |                     ---- binding `line` declared here
75  |             if let Some((key, value)) = line.split_once('=') {
    |                                         ^^^^ borrowed value does not live long enough
...
116 |         }
    |         - `line` dropped here while still borrowed
...
122 |         let Some(base_executable_home_path) = base_executable_home_path else {
    |                                               ------------------------- borrow later used here
```

I don't understand this, as the `line` variable yielded by the `universal_newlines` iterator should surely have the same lifetime as the `line` in my current PR branch.